SPEC.md

```diff
@@ -4002,28 +4019,28 @@ Optional whitespace is allowed between the number/expression and the suffix. For
 
 The decimal and binary units may be shortened by omitting the trailing "B". For example, "K" and "KB" are both interpreted as "kilobytes".
 
-#### Mandatory `runtime` attributes
+#### Requirements attributes
 
-The following attributes must be supported by the execution engine. The value for each of these attributes must be defined - if it is not specified by the user, then it must be set to the specified default value. 
+The following attributes must be supported by the execution engine. The value for each of these attributes must be defined - if it is not specified by the user, then it must be set to the specified default value.
 
 ##### `container`
 
 * Accepted types:
+    * `"*"`: This special value indicates that the runtime engine may use any POSIX-compliant operating environment it wishes to execute the task, whether that be a container or directly in the host environment.
     * `String`: A single container URI.
     * `Array[String]`: An array of container URIs.
+* Default value: `"*"`
 * Alias: `docker`
 
-The `container` key accepts a URI string that describes a location where the execution engine can attempt to retrieve a container image to execute the task.
-
-The user is strongly suggested to specify a `container` for every task. There is no default value for `container`. If `container` is not specified, the execution behavior is determined by the execution engine. Typically, the task is simply executed in the host environment. 
+The `container` attribute accepts a URI string that describes a container resource the execution engine can use when executing the task.
 
-🗑 The ability to omit `container` is deprecated. In WDL 2.0, `container` will be required.
+The user is strongly suggested to specify a `container` for every task. If `container` is not specified, or is specified with the special `"*"` value, the execution behavior is determined by the execution engine. A task that depends on the engine to determine the execution environment should be careful to only use built-in Bash operations and tools specified as mandatory by the [POSIX standard](https://unix.stackexchange.com/questions/228888/what-are-the-posix-mandatory-utilities).
 
-The format of a container URI string is `protocol://location`, where protocol is one of the protocols supported by the execution engine. Execution engines must, at a minimum, support the `docker://` protocol, and if no protocol is specified, it is assumed to be `docker://`. An execution engine should ignore any URI with a protocol it does not support.
+The format of a container URI is `protocol://location`, where `protocol` is one of the protocols supported by the execution engine. Execution engines must, at a minimum, support the `docker` protocol. If only `location` is specified, the protocol is assumed to be `docker`. An execution engine should ignore any URI with a protocol it does not support.
```

Comment:
`An execution engine should ignore any URI with a protocol it does not support.`

In almost all cases I suspect this will lead to a workflow failure. Should we call that out here? The phrase also leads to the question supporting arrays of URIs where the engine selects the container based on the protocol it supports. This might be nice for a future version of the language but I think here we need to say there will almost certainly be a failure.

Because the language doesn't have any constructs for saying how to authenticate with the repository, there is also the assumption that the engine will do this based on some configuration or roles (or fail if it doesn't know how to do this).

Reply:
Good point. If no value is specified then the default is `*`, but I don't think it's a good idea to fall back to the default if the user does specify value(s) but none are valid/supported. I made this distinction explicit.

I think the second part of your comment is about private Docker repositories - yes, it's up to the engine to specify the mechanism for how credentials are provided.